File: Lab3/F3.cs
using System;

namespace Lab2 {
    public class Function3 : Function {
        Vector o = new Vector(Lab2.N);
        Matrix mo = new Matrix(Lab2.N, Lab2.N);
        Matrix ms = new Matrix(Lab2.N, Lab2.N);
        Matrix mt = new Matrix(Lab2.N, Lab2.N);
        public Function3(FillType type) : base(type) { }
        public Function3(int num) : base(num) { }

        public override void FillWithRandom() {
            o.FillWithRandom();
            mo.FillWithRandom();
            ms.FillWithRandom();
            mt.FillWithRandom();
        }

        public override void FillWithNumber(int num = 0) {
            o.FillWithNumber(num);
            mo.FillWithNumber(num);
            ms.FillWithNumber(num);
            mt.FillWithNumber(num);
        }

        public override void FillWithInput() {
            Console.WriteLine("Enter O:");
            o.FillWithInput();
            Console.WriteLine("Enter MO:");
            mo.FillWithInput();
            Console.WriteLine("Enter MS:");
            ms.FillWithInput();
            Console.WriteLine("Enter MT:");
            mt.FillWithInput();
        }

        public override void Run() {
            // S = Sort(O * MO) * (MS * MT) 
            // lock (Lab3.Locker) {
                // Thread.Sleep(1000);
                Console.WriteLine("Function 3 start");
                var result = (o * mo).Sort() * (ms * mt);
                Console.WriteLine("Function 3 result:\n" + result);
                Console.WriteLine("Function 3 finished\n");
            // }
        }
    }
}

File: Lab5/ConsoleApplication1/F3.cpp
#include "F3.h"

F3::F3(int N) {
	this->N = N;
}

Vector* F3::getResult() {
	return result;
}

// F3  3.21  S = sort(O * MO) * (MS * MT)

void F3::run() {
	cout << "Task 3 start\n";
	Vector *O = new Vector(N);
	Matrix *MO = new Matrix(N), *MS = new Matrix(N), *MT = new Matrix(N);
	result = MS->multiply(MT)->multiply(MO->multiply(O)->sort());
	cout << "Thread 3 result:\n" << endl << result->toString() << endl;
	cout << "Task 3 end\n";
	delete O;
	delete MO;
	delete MS;
	delete MT;
}